ac-ffmpeg/src/codec/video/filter.c

```diff
@@ -0,0 +1,139 @@
+        codec_params->width, codec_params->height, codec_params->format,
+        tb_num, tb_den,
+        codec_params->sample_aspect_ratio.num, codec_params->sample_aspect_ratio.den);
+    if (avfilter_graph_create_filter(&buffersrc_ctx, buffersrc, "in", args, NULL, filter_graph) < 0) {
+        av_log(NULL, AV_LOG_ERROR, "Cannot create buffer source\n");
+        return NULL;
+    }
+    if (buffersrc_ctx == NULL) {
+        av_log(NULL, AV_LOG_ERROR, "Buffer source initialisation failed\n");
+        return NULL;
+    }
+
+    return buffersrc_ctx;
+}
+
+AVFilterContext* ffw_filtersink_new(AVFilterGraph* filter_graph) {
+    AVFilterContext* buffersink_ctx;
+    const AVFilter* buffersink = avfilter_get_by_name("buffersink");
+    enum AVPixelFormat pix_fmts[] = { AV_PIX_FMT_YUV422P, AV_PIX_FMT_YUV420P, AV_PIX_FMT_NONE };
+
+    /* buffer video sink: to terminate the filter chain. */
+    int ret = avfilter_graph_create_filter(&buffersink_ctx, buffersink, "out", NULL, NULL, filter_graph);
+    if (ret < 0) {
+        av_log(NULL, AV_LOG_ERROR, "Cannot create buffer sink\n");
+        return NULL;
+    }
+    if (buffersink_ctx == NULL) {
+        av_log(NULL, AV_LOG_ERROR, "Buffer sink initialisation failed\n");
+        return NULL;
+    }
+
+    ret = av_opt_set_int_list(buffersink_ctx, "pix_fmts", pix_fmts, AV_PIX_FMT_NONE, AV_OPT_SEARCH_CHILDREN);
+    if (ret < 0) {
+        av_log(NULL, AV_LOG_ERROR, "Cannot set output pixel format\n");
+        return NULL;
+    }
+
+    return buffersink_ctx;
+}
+
+int ffw_filtergraph_init(AVFilterGraph* filter_graph,
+    AVFilterContext* buffersrc_ctx, AVFilterContext* buffersink_ctx,
+    const char* filters_descr) {
+    int ret = 0;
+    AVFilterInOut* outputs = avfilter_inout_alloc();
+    AVFilterInOut* inputs = avfilter_inout_alloc();
+
+    /*
+     * Set the endpoints for the filter graph. The filter_graph will
+     * be linked to the graph described by filters_descr.
+     */
+
+     /*
+      * The buffer source output must be connected to the input pad of
+      * the first filter described by filters_descr; since the first
+      * filter input label is not specified, it is set to "in" by
+      * default.
+      */
+    outputs->name = av_strdup("in");
+    outputs->filter_ctx = buffersrc_ctx;
+    outputs->pad_idx = 0;
+    outputs->next = NULL;
+
+    /*
+     * The buffer sink input must be connected to the output pad of
+     * the last filter described by filters_descr; since the last
+     * filter output label is not specified, it is set to "out" by
+     * default.
+     */
+    inputs->name = av_strdup("out");
+    inputs->filter_ctx = buffersink_ctx;
+    inputs->pad_idx = 0;
+    inputs->next = NULL;
+
+    ret = avfilter_graph_parse_ptr(filter_graph, filters_descr, &inputs, &outputs, NULL);
+    if (ret < 0) {
+        return ret;
+    }
+
+    ret = avfilter_graph_config(filter_graph, NULL);
+    if (ret < 0) {
+        return ret;
+    }
+
+    return ret;
```

Comment:
Based on the FFmpeg documentation and the example [here](http://ffmpeg.org/doxygen/5.1/filtering_video_8c-example.html#_a32), the `inputs` and `outputs` should be deallocated.

```suggestion
    if ((ret = avfilter_graph_parse_ptr(filter_graph, filters_descr, &inputs, &outputs, NULL)) < 0) {
        goto end;
    }

    if ((ret = avfilter_graph_config(filter_graph, NULL)) < 0) {
        goto end;
    }

end:
    avfilter_inout_free(&inputs);
    avfilter_inout_free(&outputs);

    return ret;
```